### DISCUSSION TITLE
Require Images Use Checksums

### DISCUSSION THREAD

**Discussion by pkgajulapalli** [kyverno/policies#889] [Q&A] [answered]
Is possible that kyverno automatically identifies the sha of the given image and check if it is whitelisted instead of forcing users to specify the sha digest?

This will be helpful especially when the images are generated from helm charts.

**Comment by chipzoller** [comment]
Not quite sure what you're asking. You're referring to [this policy](https://kyverno.io/policies/other/require-image-checksum/require-image-checksum/), I presume, which only requires that an image be specified using its digest and not tag. If the digest isn't a valid digest, no image will be pulled. If you want to only allow digests from a whitelist, you can certainly do that in a separate rule/policy.

**Reply by chipzoller** [reply]
Converted to discussion.

**Reply by pkgajulapalli** [reply]
For example, if any deployment is using image with a tag instead of a digest, the policy blocks the deployment because the digest is not specified. Instead, if the digest of the image of that tag is computed and if that didn't match with what has been specified in the policy, then it will be easier to manage the images with tags itself in the deployment while ensuring only whitelisted SHAs are deployed in the cluster.

**Reply by chipzoller** [reply]
Yes, that's certainly possible with a custom policy but it wasn't the goal of the policy which you referenced.

**Reply by pkgajulapalli** [reply]
is there any reference to write such custom policy where image digest is computed & verified with a list of whitelisted SHAs?

**Reply by chipzoller** [reply]
Not that I know of, but here are some resources to help:

https://kyverno.io/docs/writing-policies/variables/#variables-from-container-images
https://kyverno.io/docs/writing-policies/external-data-sources/#variables-from-configmaps

**Reply by pkgajulapalli** [reply]
i don't think these solutions are sufficient to compute the image digest on the fly. i'll explore for more options.

**Reply by chipzoller** [reply]
They are because Kyverno will automatically look up the digest for a tag and make that available in the variable. You can then take that variable and compare it against your whitelist which you store in, for example, a central ConfigMap.

**Reply by pkgajulapalli** [reply, marked as answer]
@chipzoller I came across https://kyverno.io/policies/other/resolve-image-to-digest/resolve-image-to-digest/ policy which can get the digest from image tag. This solved my usecase.
